projects/feray/feray/data_structures.py

```diff
@@ -0,0 +1,59 @@
+from __future__ import annotations
+
+from dataclasses import dataclass, field
+from enum import Enum
+from typing import (
+    Any,
+    Callable,
+    Dict,
+    Generic,
+    Hashable,
+    List,
+    NewType,
+    Optional,
+    TypeVar,
+)
+
+DataFrameType = TypeVar("DataFrameType")
+S = TypeVar("S", bound=Hashable)
+
+RowId = NewType("RowId", str)
+CodeVersion = NewType("CodeVersion", str)
+DataVersion = NewType("DataVersion", str)
+StorageId = NewType("StorageId", str)
+MerkleHash = NewType("MerkleHash", str)
+
+
+class ValidationStatus(Enum):
+    PENDING = "pending"
+    HUMAN_VALIDATED = "human_validated"
+    LOCKED = "locked"
+
+
+@dataclass(frozen=True)
+class Provenance:
+    """Stores the metadata for a single, immutable version of a feature asset."""
+
+    code_version: CodeVersion
+    data_version: DataVersion
+    parent_hash: Optional[MerkleHash]
```

Comment:
nitpick - if the package is targeting python 3.12+ should the union-operator typing be preferred? Like `MerkleHash | None`? If it's still not determined whether the package will support <3.10 it makes sense to continue to use Optional.